utils/profile.ts

```diff
@@ -15,40 +15,113 @@ import {
+  taskEither
+} from "fp-ts/lib/TaskEither";
 import { right } from "fp-ts/lib/Either";
 import { identity } from "io-ts";
+
+import { Task } from "fp-ts/lib/Task";
+import { ServicesPreferencesModeEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/ServicesPreferencesMode";
+import {
+  makeServicesPreferencesDocumentId,
+  ServicesPreferencesModel
+} from "@pagopa/io-functions-commons/dist/src/models/service_preference";
+import { NonNegativeInteger } from "@pagopa/io-functions-commons/node_modules/@pagopa/ts-commons/lib/numbers";
+import { CosmosErrors } from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model";
 import {
+  ResponseErrorInternal,
   IResponseErrorForbiddenNotAuthorizedForRecipient,
   IResponseErrorNotFound,
   IResponseSuccessJson,
   ResponseErrorForbiddenNotAuthorizedForRecipient,
   ResponseErrorFromValidationErrors,
   ResponseErrorNotFound,
-  ResponseSuccessJson
-} from "italia-ts-commons/lib/responses";
-import { Task } from "fp-ts/lib/Task";
-import { canWriteMessage } from "../CreateMessage/handler";
+  ResponseSuccessJson,
+  IResponseErrorInternal
+} from "@pagopa/ts-commons/lib/responses";
 import { GetLimitedProfileByPOSTPayload } from "../generated/definitions/GetLimitedProfileByPOSTPayload";
+import { canWriteMessage } from "../CreateMessage/handler";
+
+// Map an error when an unexpected value is passed
+interface IUnexpectedValue {
+  readonly kind: "UNEXPECTED_VALUE";
+  readonly value: unknown;
+}
+/**
+ * Creates a IUnexpectedValue error object
+ * value is defined as never so the function can be used for exhaustive checks
+ *
+ * @param value the unexpected value
+ * @returns a formatted IUnexpectedValue error
+ */
+const unexpectedValue = (value: never): IUnexpectedValue => ({
+  kind: "UNEXPECTED_VALUE",
+  value
+});
 
 /**
  * Whether the sender service is allowed to send
  * messages to the user identified by this profile
+ * which servicesPreferencesSettings.mode is LEGACY
  */
-// eslint-disable-next-line prefer-arrow/prefer-arrow-functions
-export function isSenderAllowed(
+export const isSenderAllowedLegacy = (
   blockedInboxOrChannels:
     | RetrievedProfile["blockedInboxOrChannels"]
     | undefined,
   serviceId: ServiceId
-): boolean {
-  return (
+): TaskEither<never, boolean> =>
+  taskEither.of(
     blockedInboxOrChannels === undefined ||
-    blockedInboxOrChannels[serviceId] === undefined ||
-    blockedInboxOrChannels[serviceId].indexOf(BlockedInboxOrChannelEnum.INBOX) <
-      0
+      blockedInboxOrChannels[serviceId] === undefined ||
+      blockedInboxOrChannels[serviceId].indexOf(
+        BlockedInboxOrChannelEnum.INBOX
+      ) < 0
   );
-}
+
+/**
+ * Whether the sender service is allowed to send
+ * messages to the user identified by this profile
+ * which servicesPreferencesSettings.mode is NOT LEGACY
+ */
+export const isSenderAllowed = (
+  servicesPreferencesModel: ServicesPreferencesModel,
+  serviceId: ServiceId,
+  profileId: FiscalCode,
```

Comment:
We never use profileId as variable name for the user fiscal code. Can we use `fiscalCode` as usual?
```suggestion
  fiscalCode: FiscalCode,
```